### PR TITLE
check for exactly one NextProtocol record in the response

### PR DESCRIPTION
**Pull request by folkertdev**
the spec notes

> The NTS Next Protocol Negotiation record has a Record Type number of 1. It MUST occur exactly once in every NTS-KE request and response.

we did not check this requirement

### CODE REVIEW

File: ntpd/src/daemon/keyexchange.rs [hunk 1]
```diff
@@ -840,15 +840,23 @@ mod tests {
         let records = vec![NtsRecord::EndOfMessage];
         let result = send_records_to_server(records).await;
 
-        assert!(result.is_ok());
+        assert!(matches!(result, Err(KeyExchangeError::NoValidProtocol)));
     }
 
     #[tokio::test]
-    async fn double_end_of_message() {
-        let records = vec![NtsRecord::EndOfMessage, NtsRecord::EndOfMessage];
```
Comment (folkertdev) [inline ntpd/src/daemon/keyexchange.rs:848]: this test is flaky, and fixing that is hard in this PR

File: ntp-proto/src/nts_record.rs [hunk 2]
```diff
@@ -905,27 +919,54 @@ impl KeyExchangeResultDecoder {
...
                     None => Break(Err(NoValidProtocol)),
-                    Some(_) => Continue(state),
+                    Some(protocol) => {
+                        // The NTS Next Protocol Negotiation record [..] MUST occur exactly once in every NTS-KE request and response.
+                        match state.protocol {
+                            None => {
+                                state.protocol = Some(protocol);
+                                Continue(state)
+                            }
+                            Some(_) => Break(Err(KeyExchangeError::BadRequest)),
```
Comment (davidv1992) [inline ntp-proto/src/nts_record.rs:931]: BadRequest seems a weird description for an error which is essentially "server did not give valid response"

**Review by davidv1992 — COMMENTED** [review]
Few minor comments regarding choice of errors, but otherwise looks ok to me, though I want @squell's opinion whether this interferes with any of the previous pool work.

File: ntp-proto/src/nts_record.rs [hunk 3]
```diff
@@ -905,27 +919,54 @@ impl KeyExchangeResultDecoder {
...
+                        }
+                    }
                 }
             }
             AeadAlgorithm { algorithm_ids, .. } => {
+                // it MUST include at most one
+                let algorithm_id = match algorithm_ids[..] {
+                    [] => return Break(Err(NoValidAlgorithm)),
+                    [algorithm_id] => algorithm_id,
+                    _ => return Break(Err(BadRequest)),
```
Comment (davidv1992) [inline ntp-proto/src/nts_record.rs:941]: BadRequest seems a weird description for an error which is essentially "server did not give valid response"

**Comment by folkertdev** [comment]
fair point. None of the existing errors seemed fitting, so I added a `BadResponse` error for this case where the response is syntactically valid but semantically invalid.

**Review by squell — APPROVED** [review]
Already had a discussion about this with @folkertdev about these changes last week and they look fine to me.

**Pull request merged**